core/server/master/src/main/java/alluxio/master/file/meta/xattr/ExtendedAttribute.java

```diff
@@ -0,0 +1,75 @@
+package alluxio.master.file.meta.xattr;
+
+import com.google.protobuf.ByteString;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * This class defines an interface for implementing an extended attribute which can be stored on
+ * inodes.
+ *
+ * Attributes are stored on inodes as a simple array of bytes ({@link ByteString}) and can be
+ * the implementation for encoding/decoding
+ * @param <T>
+ */
+public interface ExtendedAttribute<T> {
+
+  PersistenceStateAttribute PERSISTENCE_STATE = new PersistenceStateAttribute();
+
+  /**
+   * @return the full attribute with namespace and identifier
+   */
+  String getName();
+
+  /**
+   * @return the namespace of the attribute lies within
+   */
+  String getNamespace();
+
+  /**
+   * @return the identifier within the namespace of the attribute
+   */
+  String getIdentifier();
+
+  /**
+   * @return the number of bytes it takes to encode a single object
+   */
+  int getEncodedSize();
+
+  /**
+   * Encode a single object into a byte string.
+   *
+   * @param object the object to encode
+   * @return the byte representation of the object
+   */
+  ByteString encode(T object);
+
+  /**
+   * Decode an object from a single byte string.
+   *
+   * @param bytes the bytes to decode into a single instance of
+   * @return the instance of the decoded object
+   * @throws IOException if the size of the byte string isn't equal to the encoding length
+   */
+  T decode(ByteString bytes) throws IOException;
+
+  /**
+   * Encode a list of objects into a byte string.
+   *
+   * @param objects the objects to encode
+   * @return a bytestring with a size equal to {@link #getEncodedSize()} * {@code objects.size()}
+   */
+  ByteString multiEncode(List<T> objects);
```

Comment:
Do we need to differentiate between single encode/decode and multi encode/decode? As a user I would assume just getting the correct value from extended attributes by calling ExtendedAttribute.MY_ATTRIBUTE.decode(some_bytes), regardless of the attribute value being a list or not.

Reply:
Actually can we require attribute values to be a protobuf type? That will make it consistent with everywhere else and save the work to write the decode/encode function.

Reply:
In my opinion, it is useful to have an explicit implementation for encode/decode because we can get tight control over exactly how many bytes are stored when an object is serialized. I don't think these attributes are intended for complex values, but I think having tighter control over the individual bytes is beneficial.

Additionally, there's nothing stopping someone from adding a proto message and simply encoding an object as the proto message and then calling `toBytes()` for encode and then using `Message.parseFrom(byte[])` in the decode function.

I tested using a simple message 

```proto
enum PersistencePState {
  NOT_PERSISTED = 1;
  TO_BE_PERSISTED = 2;
  PERSISTED = 3;
  LOST = 4;
}

message PersistenceStateAttribute {
  required PersistencePState state = 1;
}
```

In this example, when on the wire, the field `state` requires two bytes, where in fact, it should only need 1. If the field is `repeated` instead of required, the usage is still two bytes/value

For more advanced cases, we can utilize proto, but if we can encode in a smaller number of bytes than proto can, is there a reason we can't implement our own?

Lastly, I implemented multiEncode because it's possible you want to store an array/list of some kind of value as opposed to a single one. The caveat is that it requires knowledge of the size of the message before (that's why the encoding size is a parameter in the constructor).

I can see the tradeoffs of ease of use vs better memory utilization. I think we just need to decide between them. My vote is to allow our own encoding, but still support proto (the way it is now). Is there a downside to supporting both? Also, I gave persistence state as more of an example implementation than something that actually needs to be committed. So this doesn't have to necessarily be the final implementation.